src/dispatcher.rs

```diff
@@ -61,41 +64,50 @@ impl<C> CommandDispatcher<C> {
     ///
     /// Unicode characters are currently not supported. This may be fixed in the future.
     pub fn dispatch(&self, ctx: &mut C, command: &str) -> bool {
-        let parsed = Self::parse_into_arguments(command);
+        // let parsed = Self::parse_into_arguments(command);
 
         let mut current_node = self.root;
 
-        for argument in &parsed {
+        let mut input = Input::new(command);
+
+        while !input.empty() {
             // try to find a node satisfying the argument
             let node = &self.nodes[current_node.0];
-
+            
             // TODO: optimize linear search using a hash-array mapped trie
-            if let Some(next) = node.next.iter().find(|next| {
+            if let Some((next, next_input)) = node.next.iter().filter_map(|next| {
                 let kind = &self.nodes[next.0].kind;
+                let mut input = input.clone();
```

Comment:
Shouldn't the input be cloned outside of the loop, rather than inside it? Right now it's getting reset every iteration.